Icon A11y Guide (#3595)
cerner/terra-core: packages/terra-core-docs/src/terra-dev-site/doc/icon/AccessabilityGuide.2.doc.mdx

import { Notice } from "@cerner/terra-docs";
import Whitespace from "../../common/layout-helpers/Whitespace";
import { Badge } from 'terra-icon/package.json?dev-site-package';

<!-- Examples -->
import DecorativeIconExample from './example/DecorativeIconExample?dev-site-example';
import GroupOfIconsExample from './example/GroupOfIconsExample?dev-site-example';
import InformativeIconExamples from './example/InformativeIconExamples?dev-site-example';
import FunctionalIconExample from './example/FunctionalIconExample?dev-site-example';

<!-- Icons used in the examples, but also I want to use them in this doc. -->

> import IconManufacturer from 'terra-icon/lib/icon/IconManufacturer';

Reviewer: @eawww is correct. This icon is reserved exclusively for CE Marking, particularly disclosing the Electronic Instructions for Use (commonly abbreviated as eIFU) and Solution Labeling requirements. You are free to use it if you construct the full unit/fragment and use it correctly:  https://wiki.cerner.com/display/UserExperience/About+Solution+Dialog
and https://github.cerner.com/orion/orion-application/tree/master/src/about-box 

Otherwise using this icon should be avoided, as it would be suggesting others can use it for any purpose as well.

> import IconManufacturer from 'terra-icon/lib/icon/IconManufacturer';

Author: I will remove all CE icons and use different examples.

<Whitespace />

## Accessibility Considerations:

### Meaningful Icons

#### Informative Icons

Icons that convey information **must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.

<InformativeIconExamples/>

In this example the factory icon (<IconManufacturer/>) and announcement icon (<IconAnnouncement/>) are both informative. The factory icon has the meaning that _this text is an address_. The announcement icon has the meaning that _this text is an announcement_. However, neither icon conveys that meaning to users who cannot see them. Therefore, both icons code must include `title="<alternative text>"` to adequately convey their meaning to a screen reader user. The factory icon should have `title="Manufactured at"` and the announcement icon should have `title="Announcement"`. With these text alternatives added, screen reader users will hear _"&lt;Text Alternative&gt; &lt;Text&gt_":

<p><em>"Manufactured at 2800 Rock Creek Pkwy Kansas City, MO 64117"</em></p>

> <p><em>"Announcement Facilities will closed on Monday"</em></p>

Reviewer 2: Genuinely curious: Would this read to a screen reader user as 'the announcement facilities will be closed' or would it be understood as 'the text that follows the word "announcement" is the announcement'?

Author: What is read will slightly differ depending on the screen reader, but on macOS VoiceOver it will be closer to: _Image Announcement. Facilities will closed on Monday._

#### Functional Icons

> Icons that perform some operation, like an icon that is part of a hyperlink or a button, are said to be functional icons. Functional icons, like informative icons, **must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.

Reviewer 2: Same principle as above
```suggestion
Icons that perform some operation, like an icon that is part of a hyperlink or a button, are said to be functional icons. Like informative icons, **functional icons must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.
```

<p><em>"View folder as, link, List view, link, Tile view, link, Treemap view"</em></p>

#### Groups of Icons

Sometimes icons only have meaning in the context of other icons. Here is one example where multiple icons are used to create a single five-star rating system:

<GroupOfIconsExample/>

One technique to make groups of images accessible is shown above - to give the first image in the group a meaning but mark the other images as presentational (so that they will be ignored by screen readers). This technique results in the user hearing:

_"Rating: Three and a half stars out of five"_

#### Icons Containing Text

Terra does not provide Icons containig text, but you might encounter such a scenario. One example found commonly is a company logo that includes the company's name in the logo graphic. For more information see https://www.w3.org/WAI/tutorials/images/textual/.

### Decorative Use of Icons

In decorative use, an icon provides no additional information or context. Meaning that, **without the decorative icon, the user would still know the meaning and purpose of content**.

<DecorativeIconExample/>

> The factory icon (<IconManufacturer/>) above is purely decorative. It provides an extra visual cue but is not necessary for the user to understand that the text, _2800 Rock Creek Pkwy…_, is the address of the manufacturer.

Reviewer 2: CE marking icon

Since the icon is decorative, its code must set `title=""` to ensure screen readers do not announce the icon's presence. If the title is left out, then the user will hear "Image Manufactured at …". If the title is set to `title="Manufactured at"` like in the Informative example above, then the user would hear "Manufactured at Manufactured at…". Neither of these are correct. The correct choice is to set `title=""` so that the user hears:

<p><em>"Manufactured at 2800 Rock Creek Pkwy Kansas City, MO 64117"</em></p>

### Code Considerations

- Appropriately code each icon as being meaningful or decorative. Ensure each meaningful icon has alternative text provided by the content creator.
- Do not overwrite theme colors unless UX provides custom colors. All custom colors must pass WCAG 2.1 color accessibility requirements.

<Whitespace />

### Content Considerations

> - Specify **every** icon instance as meaningful or decorative. Provide developers with clear and concise alternative text for meaningful icons that accurately reflect the icon's meaning or purpose.

Reviewer 2: To clearly distinguish between the bolded main points in this document, "every" could be in italics instead,

Author: Will change.

<Whitespace />

## Accessibility Considerations:

### Meaningful Icons

#### Informative Icons

Icons that convey information **must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.

<InformativeIconExamples/>

> In this example the factory icon (<IconManufacturer/>) and announcement icon (<IconAnnouncement/>) are both informative. The factory icon has the meaning that _this text is an address_. The announcement icon has the meaning that _this text is an announcement_. However, neither icon conveys that meaning to users who cannot see them. Therefore, both icons code must include `title="<alternative text>"` to adequately convey their meaning to a screen reader user. The factory icon should have `title="Manufactured by"` and the announcement icon should have `title="Announcement"`. With these text alturnatives added, screen reader users will hear _"&lt;Text Alternative&gt; &lt;Text&gt_":

Reviewer 3: ```suggestion
In this example the factory icon (<IconManufacturer/>) and announcement icon (<IconAnnouncement/>) are both informative. The factory icon has the meaning that _this text is an address_. The announcement icon has the meaning that _this text is an announcement_. However, neither icon conveys that meaning to users who cannot see them. Therefore, both icons code must include `title="<alternative text>"` to adequately convey their meaning to a screen reader user. The factory icon should have `title="Manufactured by"` and the announcement icon should have `title="Announcement"`. With these text alternatives added, screen reader users will hear _"&lt;Text Alternative&gt; &lt;Text&gt_":
```

<p><em>"Manufactured by 2800 Rock Creek Pkwy Kansas City, MO 64117"</em></p>
<p><em>"Announcement Facilities will closed on Monday"</em></p>

#### Functional Icons

> Icons that perform some operation, like an icon taht is part of a hyperlink or a button, are said to be functional icons. Functional icons, like informative icons, **must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.

Reviewer 3: ```suggestion
Icons that perform some operation, like an icon that is part of a hyperlink or a button, are said to be functional icons. Functional icons, like informative icons, **must have alternative text in code** to ensure that assistive technologies pass the icon's meaning to the user.
```

<FunctionalIconExample/>

In the example above, links to view a folder in three different ways are given: as a list (<IconList/>), as tiles (<IconTile/>), and as a tree map (<IconTreemap/>). With the title provided to each icon, the user will hear:

<p><em>"View folder as, link, List view, link, Tile view, link, Treemap view"</em></p>

#### Groups of Icons

Sometimes icons only have meaning in the context of other icons. Here is one example where multiple icons are used to create a single five-star rating system:

<GroupOfIconsExample/>

> One technique to make groups of images accessible is show above - to give the first image in the group a meaning but mark the other images as presentationsl (so that they will be ingored by screen readers). This technique results in the user hearing:

Reviewer 3: ```suggestion
One technique to make groups of images accessible is shown above - to give the first image in the group a meaning but mark the other images as presentations (so that they will be ignored by screen readers). This technique results in the user hearing:
```

Author: Fixed in 064fb3a94e4db859a6f2f8f49f8777aca0a2da3c. I couldn't use your suggestion because I wanted to say _presentational_ instead of _presentations_.

> One technique to make groups of images accessible is show above - to give the first image in the group a meaning but mark the other images as presentationsl (so that they will be ingored by screen readers). This technique results in the user hearing:

Author: ```suggestion
One technique to make groups of images accessible is shown above - to give the first image in the group a meaning but mark the other images as presentational (so that they will be ignored by screen readers). This technique results in the user hearing:
```

_"Rating: Three and a half stars out of five"_

#### Icons Containing Text

Terra does not provide Icons containig text, but you might encounter such a scenario. One example found commonly is a company logo that includes the company's name in the logo graphic. For more information see https://www.w3.org/WAI/tutorials/images/textual/.

### Decorative Use of Icons

In decorative use, an icon provides no additional information or context. Meaning that, **without the decorative icon, the user would still know the meaning and purpose of content**.

<DecorativeIconExample/>

> The factory icon (<IconManufacturer/>) above is purely decorative. It provides an extra visual queue but is not nessicary for the user to understand that the text, _2800 Rock Creek Pkwy…_, is the address of the manufacturer.

Reviewer 3: ```suggestion
The factory icon (<IconManufacturer/>) above is purely decorative. It provides an extra visual queue but is not necessary for the user to understand that the text, _2800 Rock Creek Pkwy…_, is the address of the manufacturer.
```

<Whitespace />

## Support Compliance

Terra is committed to ensuring its components provide maximum possible accessibility. However, using Terra components will not automatically make your product accessible. At most, consuming this component accounts for 33% of a product’s accessibility — the rest is up to development teams to implement.

This component has **[known / no known]** accessibility failures. If you find accessibility bugs or gaps, please [report these issues on Github](https://github.com/cerner/terra-core/issues/new/choose).

#### Success Criteria

_Icon **must** meet the following success criteria:_

- [**1.3.1 Info and Relationships**](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html). - The intent of this Success Criterion is to ensure that information and relationships that are implied by visual or auditory formatting are preserved when the presentation format changes. For example, the presentation format changes when the content is read by a screen reader.

> - [**0.0.0. Criteria Name**](https://www.link-to-criteria.html) - copy main paragraph text for the criteria.

Reviewer 3: I'm unable to open the https://www.link-to-criteria.html link.

Author: Good catch. That is just a placeholder. I will fix.

Author: I think the criteria on line 110 sufficient, so I removed line 111 in b6239ccb3c8cf0c60aa7a37abba0e8462b771cb1.

import { Notice } from "@cerner/terra-docs";
import Whitespace from "../../common/layout-helpers/Whitespace";
import { Badge } from 'terra-icon/package.json?dev-site-package';

<!-- Examples -->
import DecorativeIconExample from './example/DecorativeIconExample?dev-site-example';
import GroupOfIconsExample from './example/GroupOfIconsExample?dev-site-example';
import InformativeIconExamples from './example/InformativeIconExamples?dev-site-example';
import FunctionalIconExample from './example/FunctionalIconExample?dev-site-example';

<!-- Icons used in the examples, but also I want to use them in this doc. -->
import IconManufacturer from 'terra-icon/lib/icon/IconManufacturer';
import IconAnnouncement from 'terra-icon/lib/icon/IconAnnouncement';
import IconTile from 'terra-icon/lib/icon/IconTile';
import IconList from 'terra-icon/lib/icon/IconList';
import IconTreemap from 'terra-icon/lib/icon/IconTreemap';

<Badge />

# Accessibility Guide for Terra Icon

<Notice variant="important" ariaLevel="2">

> Accessibility of icons is important because icons are ubiquitous. They can convey visual meaning or be used as a decorative or supportive element in a UI design and are an essential tool for creating intuitive user experiences that are appropriate for all languages. However, **if implemented improperly or without alternative content, they will not understood everyone**.

Reviewer 3: Did you mean "they will not **_be_** understood **_by_** everyone"?

Author: Fixed in 8be2421ad11674691395463c2e5518e0dfc27aae.

</Notice>

Since icons are images, they are not inherently accessible to all users, especially those with significantly impaired vision or blindness. Therefore, interface **icons must have alternative content**.

Screen readers convey alternative content to users who have difficulty interpreting icons. Speech input devices also use alternative content to target specific icons.

It is common for interfaces to use multiples of the same icon on a screen. Development and content teams must consider this use case to ensure all users can predictably understand the icon's purpose.

**Users expect identical icons to perform the same function**. Similarly, screen reader users expect their software to identify identical icons consistently. Therefore, failure to provide consistent identification (visual or alternative content) can cause problems for all users.

> In addition, people who use only their voice to interact with technology also rely upon consistent alternative content behind the icons to ensure they can interact with controls.

Reviewer 3: Should these be bullet pointed or combine in a paragraph? They currently look like individual sentences separated by a line.

![Screen Shot 2022-03-10 at 1 52 17 PM](https://user-images.githubusercontent.com/22618655/157744923-8d5bdad3-4565-477a-a5a5-6f5f21960d4d.png)

Author: I'll make them one paragraph.

Author: Fixed in 8be2421ad11674691395463c2e5518e0dfc27aae.
